dali/pipeline/operator/builtin/external_source.h

```diff
@@ -158,7 +159,7 @@ class ExternalSource : public Operator<Backend> {
       data = tl_data_.GetEmpty();
     }
 
-    data.front()->Copy(tl, 0);
+    data.front()->Copy(tl, stream);
```

Comment:
It sucks. No!!!
The external source keeps data on CPU now. With your change you will have round trip - source from GPU, copy to CPU inside SetDataSource, copy to GPU inside RunImpl. You need to extend external_source.h to be able to store data on the GPU as well.

Reply:
Fixed

Reply:
I'm not sure if this is what we wants. I can imagine a GPU ExternalSource that accepts CPU input and the opposite. @klecki  what do you think?

Reply:
With @klecki we figured out 4 scenarios:
```
Lp.    DataLocation    OpDevice
1.          CPU          CPU
2.          GPU          GPU
3.          CPU          GPU
4.          GPU          CPU
```
ad 1. Legit
ad 2. Legit
ad 3. Legit - data will be copied onto GPU by ExternalSource.
ad 4. Nope. Since in DALI once you go GPU, you never go back, I feel that this should be illegal. So if you have GPU data, don't expect to work with it on CPU op.

Remaining question is, if right now the ExternalSource will work for the 3rd scenario, I can write a test for it

Reply:
3 - it currently does and we have a test for that https://github.com/NVIDIA/DALI/blob/master/dali/test/python/test_pipeline.py#L1215
4 - it is true for the pipeline, but data on the GPU is outside the pipeline so I think we should also consider supporting it for some post-processing after inference.

Reply:
About the `4` - I believe that it's what happens now: user specifies the location of data in function invocation, it get piped to `SetExternalInputHelper` and then `SetDataSource`, where it gets copied into proper memory. The downside is, that it's a little bit obfuscated around this `SetExternalInputHelper` (we have 2 different `Backends` - storage's and operator's). I've put proper comment there.

Reply:
In the invocation, you specify where you want to get your data after the operator run. The internal storage should be the same as the input to make the copy as fast as possible as `SetDataSource` call is synchronous from the user point of view and we don't want it to last more than it is essential.

Reply:
Hmm, that's not what docs indicate:
> @param device Device of the supplied memory.

As I understood this, user specified, where the data exists and the info about where the data should be at the beginning of the pipeline is determined by operator's device in the pipeline. Am I missing something here?

Reply:
Yes, @szalpal is correct. The `device` param does indicate what kind of memory user supplies to us, so we now if we copy from CPU or GPU.
It doesn't indicate where to copy the memory, as this is the task for the Pipeline definition and the setup of ExternalSource operator which is out of scope of C API.

As you noted, the GPU (DataLocation) -> CPU (OpDevcie) may hit the inefficient scenario. It also enables users to place the ExternalSource on CPU, supply it with GPU data, and use the outputs only in GPU operators. We currently don't have any sensible way of diagnosing that, and I would like to prohibit such inefficient scenario (additional GPU->CPU->GPU copies).

If the user wants to use GPU data in CPU, unless there is a valid use-case for that, I would make them do an explicit copy on their own. API shouldn't allow to easily do bad things.

Reply:
I'm talking about Python API, as the C one should is still a secondary thing for us. So I want the Python side to be flexible and not broken in the first place.
So one thing is the device where the ExternalSource is located and the second what kind of data you supply. I think we should allow all combinations and when the data is on the GPU and ExternalSource is located on the CPU we should warn, but not break.

Reply:
So extend the ExternalSource  to house CPU and GPU storage, add some holder which will tell what kind of data should be consumed in the next iteration - CPU or GPU (maybe you can extend `data_in_tl_` and store some enum instead of bool there). Then extend the implementation with the selection of proper place holder. I don't think it it would cost much work.

Reply:
The same reasoning would apply for Python API, that GPU->CPU->GPU will be bad.

Reply:
"GPU->CPU->GPU" is bad, but "GPU->CPU" maybe useful.

Reply:
It looks OK now. The validation to (possibly) reject GPU->CPU copy should happen in Pipeline if anywhere.
I need to track the stream usage outside this function to spot potential issues at call site.